pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java

```diff
@@ -174,7 +175,8 @@ public String consumerName() {
                         .build();
 
                 ByteBuf metadataAndPayload = entry.getDataBuffer();
-
+                // increment ref-count of data and release at the end of process: so, we can get chance to call entry.release
+                entry.retain();
```

Comment:
Actually, earlier when dispatcher gives entries to [Consumer](https://github.com/yahoo/pulsar/pull/318/files/81e154d1597389f60ca3f1678449d408cb32131e#diff-95b9bf24e8d791dbf6026b1a64984900), consumer retrieves `data(byteBuf)` from it and puts bytebuf into netty-channel which decrements the ref-count by 1 (and ByteByf will be recycled if ref-count=0).

Now, as we made `EntryImpl` recyclable and it gets recycled when we call `Entry.release()` and if data ref-count =0, we recycle the `EntryImpl`. So, we `retain()` which helps later on to call [entry.release()](https://github.com/rdhabalia/pulsar/blob/81e154d1597389f60ca3f1678449d408cb32131e/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java#L201) and if entry's data ref-count is 0 then we recycle Entry and data both.

I could have avoided retain and could have recycled Entry in the [end](https://github.com/rdhabalia/pulsar/blob/81e154d1597389f60ca3f1678449d408cb32131e/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java#L201) But caller(eg. dispatcher) may not want to recycle and if caller has retain data for later use then we may break this contract. So, this scope retain and in the last try to release the entry (and it gets recycled if data ref-count=0).